File: SignalScribe/Enums/DiscardReason.cs
namespace SignalScribe.Enums;

/// <summary>Why a clip was rejected instead of becoming a transmission. Set capture-side, where the audio measurements are.</summary>
public enum DiscardReason
{
    /// <summary>Rejected, but none of the specific tests explains it.</summary>
    Unknown = 0,

    /// <summary>Barely any signal was present before the gate shut again.</summary>
    TooShort = 1,

    /// <summary>The receiver's front end was clipping, so nothing measured was real.</summary>
    FrontEndOverload = 2,

    /// <summary>A constant tone rather than speech.</summary>
    SteadyTone = 3,

    /// <summary>Almost no energy in the part of the spectrum voice occupies.</summary>
    OutsideSpeechBand = 4,

    /// <summary>Modulating far faster than anyone can speak.</summary>
    TooFastForSpeech = 5,

    /// <summary>Some voice-like audio, but not enough of it.</summary>
    NotEnoughVoice = 6,

    /// <summary>The level barely moves, so there is no syllable rhythm to find.</summary>
    NoSyllableRhythm = 7,

    /// <summary>
    /// Discrete symbol levels, so something is transmitting data — but not a mode we can name, and
    /// knowing only that a signal is digital is not grounds for creating a channel. Recording every
    /// burst on an unidentified data frequency is precisely the failure
    /// <see cref="SignalScribe.Analysis.ChannelVoiceAudit"/> exists to undo.
    /// </summary>
    DigitalNotIdentified = 8,
}

File: SignalScribe.Vue/src/lib/discardReason.ts
/** Mirrors SignalScribe.Enums.DiscardReason. */
export enum DiscardReason {
  Unknown = 0,
  TooShort = 1,
  FrontEndOverload = 2,
  SteadyTone = 3,
  OutsideSpeechBand = 4,
  TooFastForSpeech = 5,
  NotEnoughVoice = 6,
  NoSyllableRhythm = 7,
  DigitalNotIdentified = 8,
}

/** Short label for the chip, and the longer read that hangs off it on hover. */
export const DISCARD_REASONS: Record<DiscardReason, { label: string; detail: string }> = {
  [DiscardReason.Unknown]: {
    label: "unknown",
    detail: "It failed the voice check, but none of the individual tests stands out.",
  },
  [DiscardReason.TooShort]: {
    label: "too short",
    detail: "Under 200ms of signal. A click, a stray keyup, or splatter from the channel next door.",
  },
  [DiscardReason.FrontEndOverload]: {
    label: "front end overload",
    detail: "The receiver was clipping, so nothing it heard was real. Usually a transmitter close by.",
  },
  [DiscardReason.SteadyTone]: {
    label: "steady tone",
    detail: "One constant tone the whole way through. A carrier, a heterodyne, or a stuck mic.",
  },
  [DiscardReason.OutsideSpeechBand]: {
    label: "not speech",
    detail: "Hardly any energy where a voice would be. Data, noise, or a spur.",
  },
  [DiscardReason.TooFastForSpeech]: {
    label: "too fast",
    detail: "Changing far quicker than anyone can talk. Packet or APRS, most likely.",
  },
  [DiscardReason.NotEnoughVoice]: {
    label: "barely any voice",
    detail: "Sounds like someone talking, but under 800ms of it. Too little to bother transcribing.",
  },
  [DiscardReason.NoSyllableRhythm]: {
    label: "flat",
    detail: "The level hardly moves. Speech rises and falls between syllables; this does not.",
  },
  [DiscardReason.DigitalNotIdentified]: {
    label: "digital",
    detail:
      "Discrete symbol levels, so something is sending data here — but not a mode we can name, " +
      "and that alone is not enough to start recording a frequency forever.",
  },
};

export function discardLabel(reason: DiscardReason): string {
  return (DISCARD_REASONS[reason] ?? DISCARD_REASONS[DiscardReason.Unknown]).label;
}

export function discardDetail(reason: DiscardReason): string {
  return (DISCARD_REASONS[reason] ?? DISCARD_REASONS[DiscardReason.Unknown]).detail;
}
